site/docs/hive.md

```diff
@@ -17,117 +17,297 @@
+hadoopConfiguration.set(ConfigProperties.ENGINE_HIVE_ENABLED, "true"); // iceberg.engine.hive.enabled=true
 HiveCatalog catalog = new HiveCatalog(hadoopConfiguration);
 ...
 catalog.createTable(tableId, schema, spec);
 ```
 
-##### Table Property Configuration
-The property `engine.hive.enabled` needs to be set to `true` and added to the table properties when creating the Iceberg table. This can be done like so:
+#### Table property configuration
+
+Alternatively, the property `engine.hive.enabled` can be set to `true` and added to the table properties when creating the Iceberg table. 
+Here is an example of doing it programmatically:
+
 ```java
-    Map<String, String> tableProperties = new HashMap<String, String>();
-    tableProperties.put(TableProperties.ENGINE_HIVE_ENABLED, "true"); //engine.hive.enabled=true
-    catalog.createTable(tableId, schema, spec, tableProperties);
+Catalog catalog = ...;
+Map<String, String> tableProperties = Maps.newHashMap();
+tableProperties.put(TableProperties.ENGINE_HIVE_ENABLED, "true"); // engine.hive.enabled=true
+catalog.createTable(tableId, schema, spec, tableProperties);
 ```
 
-#### Query the Iceberg table via Hive
-In order to query a Hive table created by either of the HiveCatalog methods described above you need to first set a Hive configuration value like so:
-```sql
-SET iceberg.mr.catalog=hive;
+The table level configuration overwrites the global Hadoop configuration.
+
+## Iceberg and Hive catalog compatibility
+
+### Global Hive catalog
+
+From the Hive engine's perspective, there is only 1 global data catalog, which is the Hive metastore defined in the Hadoop configuration in the runtime environment.
+On contrast, Iceberg supports multiple different data catalog types such as Hive, Hadoop, AWS Glue, and also allow any custom catalog implementations.
+Users might want to read tables in anther catalog through the Hive engine, or perform cross-catalog operations like join.
+
+Iceberg handles this issue in the following way:
+
+1. All tables created by Iceberg's `HiveCatalog` with Hive engine feature enabled are automatically visible by the Hive engine.
+2. For Iceberg tables created in other catalogs, the catalog information is registered through Hadoop configuration.
+A Hive external table overlay needs to be created in the Hive metastore, 
+and the actual catalog name is recorded as a part of the overlay table properties. 
+See [CREATE EXTERNAL TABLE](#create-external-table) section for more details.
+
+### Custom Iceberg catalogs
+
+To globally register different catalogs, set the following Hadoop configurations:
+
+| Config Key                                    | Description                                            |
+| --------------------------------------------- | ------------------------------------------------------ |
+| iceberg.catalog.<catalog_name\>.type          | type of catalog: `hive`,`hadoop` or `custom`             |
```

Comment:
Is `custom` supported? I thought that if `catalog-impl` is set, it overrides the value of `type`.

Reply:
Currently the `type` must not be null. It can be any string, not necessarily `custom`. If null, the current logic will not load any catalog implementation because the catalog type is identified as `NO_CATALOG_TYPE`:

https://github.com/apache/iceberg/blob/83886c87f516afed7e625bed7f5749838217138c/mr/src/main/java/org/apache/iceberg/mr/Catalogs.java#L206-L207

https://github.com/apache/iceberg/blob/83886c87f516afed7e625bed7f5749838217138c/mr/src/main/java/org/apache/iceberg/mr/Catalogs.java#L269-L272

Reply:
@rdblue what you describe is the behavior for loading in Spark or Flink, and `type` can be `null` when `catalog-impl` is set.

`catalog-impl` does override `type` if both are set and both values are correct, but if someone sets `catalog-impl=xxxxx` and `type=hive`, it is most likely something wrong with the user input, that's why a consistent value `custom` is suggested (and also used in tests based on what I see).

I think we should probably have followed the same way in Hive to keep the experience in all engines consistent, but I did not notice this until the PR was merged, so really sorry for the inconsistency.

@lcspinter was there any concern at that time that prevented us from following the same pattern as Spark and Flink? If not, we can probably have another PR to make them consistent.

(this does not need to happen before this PR, we can merge this first and change it later).

Reply:
Yeah, we should fix this to have consistent behavior across engines.

Reply:
+1 for consistency

Reply:
@rdblue @aokolnychyi cool, I will fix it in another PR.